### DISCUSSION TITLE
Visualization

### DISCUSSION THREAD

**Discussion by DoSquared** [ultralytics/yolov5#8919] [General]
I was wondering how to visualize the architecture and intermediate steps of the detection to create a figure for a publication. 

For example something similar to this figures from Xu. et al
![image](https://user-images.githubusercontent.com/4492013/183853612-d3e7f07e-3b99-4d42-a30b-78f3ad7ae6ea.png)

**Comment by glenn-jocher** [comment]
@DaliaMahdy 👋 Hello! Thanks for asking about YOLOv5 🚀 **architecture visualization**. We've made visualizing YOLOv5 🚀 architectures super easy. There are 3 main ways:

## `model.yaml`

Each model has a corresponding yaml file that displays the model architecture. Here is YOLOv5s, defined by `yolov5s.yaml`:
https://github.com/ultralytics/yolov5/blob/1a3ecb8b386115fd22129eaf0760157b161efac7/models/yolov5s.yaml#L12-L48

## TensorBoard Graph

Simply start training a model, and then view the TensorBoard Graph for an interactive view of the model architecture. This example shows YOLOv5s viewed in our [Notebook](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) –  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

```python
# Tensorboard
%load_ext tensorboard
%tensorboard --logdir runs/train

# Train YOLOv5s on COCO128 for 3 epochs
python train.py --weights yolov5s.pt --epochs 3
```

<img width="893" alt="Screenshot 2021-04-11 at 01 10 09" src="https://user-images.githubusercontent.com/26833433/114286928-349bd600-9a63-11eb-941f-7139ee6cd602.png">

## Netron viewer

Use [https://netron.app](https://netron.app/) to view exported ONNX models:

```
python export.py --weights yolov5s.pt --include onnx --simplify
```

<img width="890" alt="Screen Shot 2022-04-29 at 11 09 23 AM" src="https://user-images.githubusercontent.com/26833433/165999628-0d6095ce-58dd-49b7-a04e-25d0995e5493.png">


Good luck 🍀 and let us know if you have any other questions!

**Reply by DoSquared** [reply]
That is very helpful! Thank you

**Comment by DoSquared** [comment]
Could you also give me advice on how to visualize detection intermediate steps

I could export the feature maps by adding --visualize on running detect.py

But I am looking for something similar to the figure below: 

![image](https://user-images.githubusercontent.com/4492013/190157946-7e4dae9e-0868-402c-9c44-c77a1c7c8d24.png)
